ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp

```diff
@@ -324,35 +330,50 @@ class FoxgloveBridge : public rclcpp::Node {
     // Create a QoS profile that only keeps the last message since we don't need to keep a history,
     // and use reliable delivery if all publishers are reliable otherwise best-effort
     bool reliable = true;
+    bool durationVolatile = true;
+    size_t depth = 1;
     for (const auto& publisher : this->get_publishers_info_by_topic(topic)) {
-      if (publisher.qos_profile().reliability() == rclcpp::ReliabilityPolicy::BestEffort) {
+      const auto& qos = publisher.qos_profile();
+      if (qos.reliability() == rclcpp::ReliabilityPolicy::BestEffort) {
         reliable = false;
-        break;
       }
+      if (qos.durability() == rclcpp::DurabilityPolicy::TransientLocal) {
+        durationVolatile = false;
+      }
+      depth += qos.depth();
     }
 
-    rclcpp::QoS qos{rclcpp::KeepLast(1)};
+    rclcpp::QoS qos{rclcpp::KeepLast(depth)};
     if (!reliable) {
       qos.best_effort();
     }
+    if (!durationVolatile) {
+      qos.transient_local();
+    }
 
     try {
       auto subscriber = this->create_generic_subscription(
         topic, datatype, qos,
-        [&](std::shared_ptr<rclcpp::SerializedMessage> msg) {
-          rosMessageHandler(channel, msg);
-        },
+        std::bind(&FoxgloveBridge::rosMessageHandler, this, channel, clientHandle, _1),
```

Comment:
Had to convert this from a lambda function to std::bind to avoid segmentation faults. It would have probably also worked with a lambda that captures `=clientHandle` (as copy).